p2p/ipld/nmt_wrapper.go

```diff
@@ -55,7 +57,11 @@ func (w *ErasuredNamespacedMerkleTree) Push(data []byte, idx rsmt2d.SquareIndex)
 	if idx.Axis+1 > uint(w.squareSize) || idx.Cell+1 > uint(w.squareSize) {
 		copy(nsID, types.ParitySharesNamespaceID)
 	} else {
-		copy(nsID, data[:types.NamespaceSize])
+		if bytes.Equal(data[:types.NamespaceSize], nsID){
+			copy(nsID, types.TailPaddingNamespaceID)
+		} else {
+			copy(nsID, data[:types.NamespaceSize])
+		}
```

Comment:
This checks if we receive a share without NamespaceID and sets a tailpadding one.

Reply:
However, I also guess this is not the final and fulfilling solution, as the given share here should probably have a namespace in it instead of zeros. It seems like if we get a non-zero block with only one share it would still be zeroed in this case and applying tail padding is wrong. For some reason, NamespaceId is just not recovered.

Reply:
this is best solution imho, as the nmt wrapper's entire purpose is to use the correct namespaces for erasure and non erasure data. It had no need to use the Tail padding namespace until #282, which should have accounted for this (my mistake!).

However, I do think a comment describing that we're using the `TailPaddingnamespaceID` would be a good idea for future readers.

Reply:
@evan-forbes, where do you want me to put the comment?

Reply:
I don't think the placement matters that much, perhaps just above the nested `if` statement, or just with the other comment. :man_shrugging:

Reply:
Still, I don't get why the tail padding namespace is not getting repaired. We set it before extending and reference the data with a namespace in it, so it is not clear why we end up setting it manually.

Reply:
ahh I see. This nmt wrapper is non-obvious and weird. It's weird because we need to include erasure namespaces in the erasure data, and nmt was built to be more general purpose. This means that when we include the namespaces in the first half of the tree ([Q0](https://github.com/lazyledger/lazyledger-specs/blob/master/specs/data_structures.md#2d-reed-solomon-encoding-scheme)), the second half of the tree will just be erasure data. Which of course means it doesn't have namespaces, and we have to add them manually. To keep nmt general purpose, we use a wrapper that adds namespaces to everything, even the data that already has a namespace. This way nmt works like normal, *and* the namespaces are included in the erasure data.

Its clear that this whole package needs more documentation, so I'm going to approve w/o the comment and I'll add something similar to the above paragraph, along with any comments in a small PR.